### DISCUSSION TITLE
Regression in connection test latency

### DISCUSSION THREAD

**Discussion by klzgrad** [SagerNet/SagerNet#437] [General]
The latency measured in the builtin URL request tests was close to raw ping numbers in 0.8-beta01 (and the few releases before), which was good because the latency used to be unknowingly several hundred milliseconds higher in older releases. In 0.8-beta02 the latency went back to being several hundred ms higher than raw ping numbers.

I can't give specific steps to reproduce or pinpoint the cause right now. Part of it is because I previously tested this latency in x86 Android Studio but the latency was also close to raw pings without the unknown overhead. Maybe the VM is too fast to represent realistic performance of an ARM mobile device where the overhead happens.

I just want to bring up this issue of latency performance optimization, which is usually overlooked for throughput performance but I think it matters more to the users in most cases because the users are mostly in interactive instead of downloading large files scenarios where latency matters more than throughput. The result from the connection test alone is not enough to locate latency introduced in the intermediate steps in a proxied connection. Need to trace the timestamps of those steps to locate where the overhead is introduced.

To give an idea what intermediate timestamps can be traced:

* 1- Time of the first "packet" in a connection from /dev/tun read or VpnService's abstraction of it.
* 2- Time of the first packet or write to the proxy protocol stack, e.g. Trojan's listening port or the API that accepts a request (measures TCP stack overhead, gVisor or otherwise)
* 3- Time of the first packet sent to remote server (measures the proxy stack overhead), I think this cannot be directly measured without help from the proxy stack.
* 4,5,6 the other direction returning.

Looks like just one overhead of the TCP stack can be traced, which makes this thing even easier. So the question is then, the connection test tool or something else should be able to measure the latency overhead introduced by VpnService and gVisor, so when the gVisor stack is changed and updated, the performance numbers give specific feedback on changes in performance.

**Comment by nekohasekai** [comment]
The connection test is always performed directly by the core without going through the VPNService (DNS resolution time is not calculated).
